File: backend/api/types/problem/problem.go
package problem

import "backend/api/types/base"

type Problem struct {
	base.BaseModel

	Title       string     `json:"title" gorm:"size:255;not null"`
	Description string     `json:"description" gorm:"type:text;not null"`
	Language    Language   `json:"language" gorm:"size:50;not null;index"`
	Difficulty  Difficulty `json:"difficulty" gorm:"size:20;not null;default:easy;index"`

	// Limites d'exécution
	TimeLimit   int `json:"time_limit_ms" gorm:"default:2000"`  // millisecondes
	MemoryLimit int `json:"memory_limit_mb" gorm:"default:256"` // mégaoctets

	// Metadata
	AuthorID *int64  `json:"author_id,omitempty" gorm:"index"`
	Points   *int32  `json:"points" gorm:"default:100"`

	// Code templates
	StarterCode  *string `json:"starter_code,omitempty" gorm:"type:text"`
	SolutionCode *string `json:"-" gorm:"type:text"` // Jamais exposé aux étudiants

	// Relations
	TestCases   []TestCase    `json:"test_cases,omitempty" gorm:"foreignKey:ProblemID"`
	ProblemSets []*ProblemSet `json:"problem_sets,omitempty" gorm:"many2many:problem_problem_sets"`
}

// TableName spécifie le nom de la table
func (Problem) TableName() string {
	return "problems"
}

// GetVisibleTestCases retourne uniquement les test cases visibles
func (p *Problem) GetVisibleTestCases() []TestCase {
	var visible []TestCase
	for _, tc := range p.TestCases {
		if !tc.Hidden {
			visible = append(visible, tc)
		}
	}
	return visible
}

// GetHiddenTestCases retourne uniquement les test cases cachés
func (p *Problem) GetHiddenTestCases() []TestCase {
	var hidden []TestCase
	for _, tc := range p.TestCases {
		if tc.Hidden {
			hidden = append(hidden, tc)
		}
	}
	return hidden
}

File: frontend/src/types/problem.ts
// Problem types (from problem.go)

import { BaseModel } from './base';
import { Difficulty, Language } from './enums';
import { TestCase, CreateTestCaseRequest } from './testcase';
import { ProblemSet } from './problemset';

export interface Problem extends BaseModel {
    title: string;
    description: string;
    language: Language;
    difficulty: Difficulty;
    time_limit_ms: number;
    memory_limit_mb: number;
    author_id: number;
    points?: number;
    starter_code?: string;
    solution_code?: string;
    test_cases?: TestCase[];
    problem_sets?: ProblemSet[];
}

// Request DTOs
export interface CreateProblemRequest {
    title: string;
    description: string;
    language: Language;
    difficulty: Difficulty;
    time_limit_ms?: number;
    memory_limit_mb?: number;
    points?: number;
    starter_code?: string;
    solution_code?: string;
    test_cases?: CreateTestCaseRequest[];
}

export interface UpdateProblemRequest {
    title?: string;
    description?: string;
    language?: Language;
    difficulty?: Difficulty;
    time_limit_ms?: number;
    memory_limit_mb?: number;
    points?: number;
    starter_code?: string;
    solution_code?: string;
}

export interface ListProblemsParams {
    page?: number;
    limit?: number;
    language?: Language;
    difficulty?: Difficulty;
    search?: string;
}

// Response DTO
export interface ProblemListItem {
    id: number;
    title: string;
    language: Language;
    difficulty: Difficulty;
    points: number;
}
